raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java

```diff
@@ -0,0 +1,137 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Checks if a trace can be finalized and emitted based on inactivity period of {@link
+ * RawSpansGroupingTransformer#groupingWindowTimeoutMs}
+ */
+public class TraceEmitPunctuator implements Punctuator {
+
+  private static final Logger logger = LoggerFactory.getLogger(TraceEmitPunctuator.class);
+
+  private double dataflowSamplingPercent = -1;
+  private final String DATAFLOW_SAMPLING_PERCENT = "dataflow.metriccollection.sampling.percent";
+  static final String TRACE_CREATION_TIME = "trace.creation.time";
+  private static final Timer spansGrouperArrivalLagTimer =
+      PlatformMetricsRegistry.registerTimer(DataflowMetricUtils.ARRIVAL_LAG, new HashMap<>());
+
+  private String key;
+  private ProcessorContext context;
+  private KeyValueStore<String, ValueAndTimestamp<RawSpans>> inflightTraceStore;
+  private KeyValueStore<String, Long> traceEmitTriggerStore;
+  private To outputTopicProducer;
+  private long groupingWindowTimeoutMs;
+  private Cancellable cancellable;
+
+  public TraceEmitPunctuator(String key, ProcessorContext context,
+      KeyValueStore<String, ValueAndTimestamp<RawSpans>> inflightTraceStore,
+      KeyValueStore<String, Long> traceEmitTriggerStore,
+      To outputTopicProducer,
+      long groupingWindowTimeoutMs) {
+    this.key = key;
+    this.context = context;
+    this.inflightTraceStore = inflightTraceStore;
+    this.traceEmitTriggerStore = traceEmitTriggerStore;
+    this.outputTopicProducer = outputTopicProducer;
+    this.groupingWindowTimeoutMs = groupingWindowTimeoutMs;
+  }
+
+  public void setCancellable(Cancellable cancellable) {
+    this.cancellable = cancellable;
+  }
+
+  /**
+   * @param timestamp correspond to current system time
+   */
+  @Override
+  public void punctuate(long timestamp) {
+    // always cancel the punctuator else it will get re-scheduled automatically
+    cancellable.cancel();
+
+    Long emitTs = traceEmitTriggerStore.delete(key);
+    if (emitTs <= timestamp) {
+      // Implies that no new spans for the trace have arrived within the last 'sessionTimeoutMs' interval
+      // so the trace can be finalized and emitted
+      ValueAndTimestamp<RawSpans> agg = inflightTraceStore.delete(key);
+      ByteBuffer traceId = null;
+      String customerId = null;
+      if (!agg.value().getRawSpans().isEmpty()) {
+        RawSpan firstSpan = agg.value().getRawSpans().get(0);
+        traceId = firstSpan.getTraceId();
+        customerId = firstSpan.getCustomerId();
+      }
+      List<RawSpan> rawSpanList = new ArrayList<>();
+      for (IndexedRecord r : agg.value().getRawSpans()) {
+        rawSpanList.add((RawSpan) r);
+      }
+
+      Timestamps timestamps = trackEndToEndLatencyTimestamps(timestamp, agg);
+      StructuredTrace trace = StructuredTraceBuilder
+          .buildStructuredTraceFromRawSpans(rawSpanList, traceId, customerId,
```

Comment:
Curious: Can we directly use `agg.value().getRawSpans()` instead of creating new `rawSpanList` in lines 90-93?

Reply:
fixed